### PR TITLE
fix some logging brokenness

### PR DESCRIPTION
**Pull request by JanJecmen**
* files were not working (everything went into the same file)
* files were written into .pir directory, if that didn't exist, nothing would be logged (c++ doesn't deal with directories I think) - one idea would be to have the directory as a debug option and create the directory from R
* files were opened for the whole logger lifetime, now use stringstreams instead and only write at the end
* contradicting options - both print into files and stdout would only print into files, now it does both
* allocator was not implemented
* use raii to clean up the logger

### CODE REVIEW

File: rir/R/rir.R [hunk 1]
```diff
@@ -20,8 +20,8 @@ rir.compile <- function(what) {
 }
 
 # optimizes given rir compiled closure
-pir.compile <- function(what, debugFlags = pir.debugFlags()) {
-    .Call("pir_compile", what, debugFlags)
+pir.compile <- function(what, debugFlags) {
+    .Call("pir_compile", what, if (missing(debugFlags)) NULL else debugFlags)
```
Comment (charig) [inline rir/R/rir.R:24]: Null or R_NilValue?

File: rir/R/rir.R [hunk 2]
```diff
@@ -32,28 +32,28 @@ pir.tests <- function() {
...
+          PrintOptimizationPasses, PrintCSSA, PrintAllocator, PrintFinalPir,
+          PrintFinalRir,
           # wants a dummy parameter at the end for technical reasons
           NULL)
 }
 
 # sets the default debug options for pir compiler
 pir.setDebugFlags <- function(debugFlags = pir.debugFlags()) {
-    .Call("pir_setDebugFlags", debugFlags)
+    invisible(.Call("pir_setDebugFlags", debugFlags))
```
Comment (charig) [inline rir/R/rir.R:56]: Why invisible is needed here?

**Review by charig — APPROVED** [review]

File: rir/R/rir.R (hunk 2 above)
Comment (JanJecmen) [reply rir/R/rir.R:56]: it's not needed, it just hides the returned NULL so that it isn't printed in the console

File: rir/R/rir.R (hunk 1 above)
Comment (JanJecmen) [reply rir/R/rir.R:24]: `NULL` in R ~ `R_NilValue` in C

Comment (charig) [reply rir/R/rir.R:24]: He, sorry I read it like c code...

**Comment by o-** [comment]
thanks a lot

**Pull request merged**